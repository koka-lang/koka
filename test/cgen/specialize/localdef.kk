// Applies a function `f` to list of increasing elements from `lo`  to `hi`
// (including both `lo`  and `hi` ).
// If `lo > hi`  the function returns the empty list.
inline fun li( lo : int, hi: int, f : int -> e a) : e list<a>
  fun enumerate( low:int, high:int, acc )
    if (low > high)
     then acc
     else enumerate(low, unsafe-decreasing(high - 1), Cons(f(high),acc))
  enumerate(lo,hi,[])

fun main()
  li(0, 10) fn(i)
    i.show
  .show.println