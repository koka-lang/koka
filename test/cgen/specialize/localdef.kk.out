["0","1","2","3","4","5","6","7","8","9","10"]
 
cgen/specialize/localdef/.lift000-li: forall<a,e> (f : (int) -> e a, low : int, high : int, acc : list<a>) -> e list<a>
cgen/specialize/localdef/.lift000-main: (low : int, high : int, acc : list<string>) -> console list<string>
cgen/specialize/localdef/.mlift000-op: forall<a,e> (acc : list<a>, f : (int) -> e a, a00.000 : int, low : int, a) -> e list<a>
cgen/specialize/localdef/li: forall<a,e> (lo : int, hi : int, f : (int) -> e a) -> e list<a>
cgen/specialize/localdef/main: () -> console ()